src/github.com/matrix-org/go-neb/services/google/google.go

```diff
@@ -0,0 +1,221 @@
+// 	Cx string `json:"cx"`
+// }
+
+type googleSearchResults struct {
+	SearchInformation struct {
+		TotalResults int64 `json:"totalResults,string"`
+	} `json:"searchInformation"`
+	Items []googleSearchResult `json:"items"`
+}
+
+type googleSearchResult struct {
+	Title       string      `json:"title"`
+	HTMLTitle   string      `json:"htmlTitle"`
+	Link        string      `json:"link"`
+	DisplayLink string      `json:"displayLink"`
+	Snippet     string      `json:"snippet"`
+	HTMLSnippet string      `json:"htmlSnippet"`
+	Mime        string      `json:"mime"`
+	FileFormat  string      `json:"fileFormat"`
+	Image       googleImage `json:"image"`
+}
+
+type googleImage struct {
+	ContextLink     string  `json:"contextLink"`
+	Height          float64 `json:"height"`
+	Width           float64 `json:"width"`
+	ByteSize        int64   `json:"byteSize"`
+	ThumbnailLink   string  `json:"thumbnailLink"`
+	ThumbnailHeight float64 `json:"thumbnailHeight"`
+	ThumbnailWidth  float64 `json:"thumbnailWidth"`
+}
+
+// Service contains the Config fields for the Google service.
+// TODO - move the google custom search engine ID in here!
+//
+// Example request:
+//   {
+//       "api_key": "AIzaSyA4FD39m9pN-hiYf2NRU9x9cOv5tekRDvM"
+//   }
+type Service struct {
+	types.DefaultService
+	// The Google API key to use when making HTTP requests to Google.
+	APIKey string `json:"api_key"`
+}
+
+// Commands supported:
+//    !google some search query without quotes
+// Responds with a suitable image into the same room as the command.
+func (s *Service) Commands(client *gomatrix.Client) []types.Command {
+	return []types.Command{
+		types.Command{
+			Path: []string{"google"},
+			Command: func(roomID, userID string, args []string) (interface{}, error) {
+				return s.cmdGoogle(client, roomID, userID, args)
+			},
+		},
+	}
+}
+
+// usageMessage returns a matrix TextMessage representation of the service usage
+func usageMessage() *gomatrix.TextMessage {
+	return &gomatrix.TextMessage{"m.notice",
+		`Usage: !google image image_search_text`}
+}
+
+func (s *Service) cmdGoogle(client *gomatrix.Client, roomID, userID string, args []string) (interface{}, error) {
+
+	if len(args) < 2 || args[0] != "image" {
+		return usageMessage(), nil
+	}
+	// Drop the search type (should currently always be "image")
+	args = args[1:]
+
+	// only 1 arg which is the text to search for.
+	querySentence := strings.Join(args, " ")
+
+	searchResult, err := s.text2imgGoogle(querySentence)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var imgURL = searchResult.Link
+	if imgURL == "" {
+		return gomatrix.TextMessage{
+			MsgType: "m.text.notice",
+			Body:    "No image found!",
+		}, nil
+	}
+
+	// FIXME -- Sometimes upload fails with a cryptic error - "msg=Upload request failed code=400 "
+	resUpload, err := client.UploadLink(imgURL)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to upload Google image to matrix: %s", err.Error())
+	}
+
+	img := searchResult.Image
```

Comment:
Careful, this isn't just taking a reference to `searchResult.Image`, it's making a copy of it (`searchResult.Image` has the type `googleImage` which is a `struct` which is a value type, see https://play.golang.org/p/44IJ0lwXXb ). This isn't a problem in this case aside from it being needlessly copied, but I suspect you intended to alias `searchResult.Image` to a shorter name for use when you inline reference `Height` and `Width`, and didn't intend to copy it.

As a general rule of thumb, don't alias value types unless you need to since otherwise you're just needlessly copying things. You can alias value types by taking pointers to them:
```go
// https://play.golang.org/p/Q-WQzD4Lw5
img := &searchResult.Image
```
If it was already a pointer, then it's a reference type and you could alias it as you are: https://play.golang.org/p/S1WmzQRN1b

That's the "how" but not the "why". You should only be aliasing variables like this if you have a good reason to, which is typically one of:
 - To improve clarity e.g `searchEngineID := query.Cx`
 - To make it shorter if using it will be extremely long e.g `id := rooms[0].Events[0].eventID`

We don't have any line length requirements in our Go style guide. We have a rough rule of thumb which is rather unfortunately based on the whim of Github: try not to make code in your PR overflow such that a horizontal scrollbar is created. In this particular case, `searchResult.Image` isn't long enough to warrant aliasing given the length of `:134`. Having fewer variables lying around in a function reduces the number of things a reader needs to hold in their head at any one time, which is a Good Thing 😄